c7n/resources/s3.py

```diff
@@ -789,6 +1029,212 @@ def process_bucket(self, func, bucket, topic, account_id, session_factory):
         return source.add(func)
 
 
+@actions.register('configure-lifecycle')
+class ConfigureLifecycle(BucketActionBase):
+    """Action that aggregates all rules with whole bucket prefix and
+    changes them to the configured policy values."""
+
+    permissions = ('s3:GetLifecycleConfiguration', 's3:PutLifecycleConfiguration', )
+
+    schema = {
+        'type': 'object',
+        'additionalProperties': True,
+        'properties': {
+            'type': {'enum': ['configure-lifecycle']},
+            'id': {'type': 'string'},
+            'prefix': {'type': 'string'},
+            'deletes_objects': {'type': 'boolean'},
+            'delete_days': {'type': 'number'},
+            'multipart_days': {'type': 'number'},
+            'ia_days': {'type': 'number'},
+            'glacier_days': {'type': 'number'},
+            'max_workers': {'type': 'number'}
+        }
+    }
+
+    cache_update = False
+
+    def process(self, buckets):
+            max_workers = self.data.get('max_workers', DEFAULT_WORKERS)
+
+            with self.executor_factory(max_workers=max_workers) as w:
+                saved_buckets = w.map(self.call_api, buckets)
+                saved_buckets = filter(None, list(saved_buckets))
+
+            if self.cache_update:
+                saved_buckets = {b['Name']: b for b in saved_buckets}
+                self.manager._cache.save(CACHE_NAME, saved_buckets)
```

Comment:
Yes. We decided to cache these values because the lifecycle API takes a long time if there are a large amount of buckets. The caching is meant to optimize a policy that uses both the filter `has-lifecycle` and the action `configure-lifecycle`,  so that the lifecycle is only read once, if both are used.

In our environment, we are clearing the cache before each run, so any changes will be reflected. The caching can be removed for the general case, but it should be noted that will result in long processing times for large environments.